src/cargo/util/config/mod.rs

```diff
@@ -174,6 +177,9 @@ pub struct Config {
     upper_case_env: HashMap<String, String>,
     /// Tracks which sources have been updated to avoid multiple updates.
     updated_sources: LazyCell<RefCell<HashSet<SourceId>>>,
+    /// Cache of credentials from configuration or credential providers.
+    /// Maps from url to credential value.
+    credential_cache: LazyCell<RefCell<HashMap<CanonicalUrl, String>>>,
```

Comment:
Since this now can potentially make 100s of requests to authenticated sparse registries, we need somewhere to cache the credential. This is especially true with credential providers, since we don't want to call it for every request.